File: src/graphs2/johnson.cpp
// Johnson APSP（CLRS §23.3）：超級源點 Bellman-Ford 求 h → 重加權 → V 次 Dijkstra
#include <iostream>
#include <queue>
#include <vector>

const long long INF = 1e18;
struct Edge { int u, v; long long w; };

int main() {
    const int n = 5;                                 // CLRS 圖 23.1
    std::vector<Edge> E = {{0,1,3},{0,2,8},{0,4,-4},{1,3,1},{1,4,7},
                           {2,1,4},{3,0,2},{3,2,-5},{4,3,6}};

    // ① 超級源點 q（編號 n）連到所有頂點，Bellman-Ford 求 h
    std::vector<Edge> Eq = E;
    for (int v = 0; v < n; ++v) Eq.push_back({n, v, 0});
    std::vector<long long> h(n + 1, 0);              // d(q,·) 初值 0 起跑
    for (int i = 0; i < n; ++i)
        for (const Edge& e : Eq)
            if (h[e.u] + e.w < h[e.v]) h[e.v] = h[e.u] + e.w;
    std::cout << "h = ";
    for (int v = 0; v < n; ++v) std::cout << h[v] << " ";
    std::cout << "\n";

    // ② 重加權 w' = w + h(u) − h(v) ≥ 0
    std::vector<std::vector<std::pair<int,long long>>> adj(n);
    for (const Edge& e : E) adj[e.u].push_back({e.v, e.w + h[e.u] - h[e.v]});

    // ③ 每個源點 Dijkstra，再還原距離
    std::cout << "D（與 Floyd-Warshall 相同）：\n";
    for (int s = 0; s < n; ++s) {
        std::vector<long long> dist(n, INF);
        using P = std::pair<long long,int>;
        std::priority_queue<P, std::vector<P>, std::greater<P>> pq;
        dist[s] = 0; pq.push({0, s});
        while (!pq.empty()) {
            auto [du, u] = pq.top(); pq.pop();
            if (du > dist[u]) continue;
            for (auto [v, w] : adj[u])
                if (dist[u] + w < dist[v]) {
                    dist[v] = dist[u] + w;
                    pq.push({dist[v], v});
                }
        }
        std::cout << "  ";
        for (int v = 0; v < n; ++v)
            std::cout << dist[v] + h[v] - h[s] << " ";   // 還原原權重距離
        std::cout << "\n";
    }
    return 0;
}

File: src/graphs2/johnson.py
# Johnson 全點對最短路徑（改寫自官方 Chapter 23/johnson.py，MIT License）：
# 超級源點 Bellman-Ford 求 h → 重新加權 w'(u,v)=w+h(u)−h(v) ≥ 0 → 每點跑 Dijkstra
import heapq

INF = float("inf")
n = 5
E = [(0,1,3),(0,2,8),(0,4,-4),(1,3,1),(1,4,7),(2,1,4),(3,0,2),(3,2,-5),(4,3,6)]

# ① 超級源點 q 連到所有頂點（權 0），Bellman-Ford 求 h
h = [0] * (n + 1)                            # 頂點 n 當 q，h 初值即 d(q,·)=0 起跑
edges_q = E + [(n, v, 0) for v in range(n)]
for _ in range(n):
    for u, v, w in edges_q:
        if h[u] + w < h[v]:
            h[v] = h[u] + w
assert all(h[u] + w >= h[v] for u, v, w in edges_q), "有負環"
print("h =", h[:n])

# ② 重新加權：w'(u,v) = w + h(u) − h(v) ≥ 0（路徑長守恆：差固定 h(u)−h(v)）
adj = {u: [] for u in range(n)}
for u, v, w in E:
    adj[u].append((v, w + h[u] - h[v]))

# ③ 每個源點跑一次 Dijkstra，再把距離「還原」回原權重
D = []
for src in range(n):
    dist = [INF] * n
    dist[src] = 0
    pq = [(0, src)]
    while pq:
        du, u = heapq.heappop(pq)
        if du > dist[u]:
            continue
        for v, w in adj[u]:
            if dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                heapq.heappush(pq, (dist[v], v))
    D.append([dist[v] + h[v] - h[src] if dist[v] < INF else INF
              for v in range(n)])            # 還原：加回 h(v) − h(src)

print("D（與 Floyd-Warshall 相同）：")
for row in D:
    print("  ", row)
